sdks/java/core/src/main/java/org/apache/beam/sdk/io/FileBasedSink.java

```diff
@@ -146,21 +159,119 @@ public String getMimeType() {
    */
   protected final WritableByteChannelFactory writableByteChannelFactory;
 
+
   /**
-   * Base filename for final output files.
+   * A naming policy for output files.
    */
-  protected final ValueProvider<String> baseOutputFilename;
+  public abstract static class FilenamePolicy
+      implements org.apache.beam.sdk.transforms.SerializableFunction<
+       FilenamePolicy.Context, String> {
+    /**
+     * Context used for generating a name based on window, pane, shard numer, and num shards.
+     * Window and pane will only be provided if windowed writes have been requested using
+     * {@link Write.Bound#withWindowedWrites}. The policy must produce unique filenames for unique
+     * Context objects.
+     */
+    public static class Context {
+      @Nullable private BoundedWindow window;
+      @Nullable private PaneInfo paneInfo;
+      private int shardNumber;
+      private int numShards;
+
+      public Context(
```

Comment:
The implication of a `Context` parameter is that it is just an optional bunch of bits that you could use or not based on your logic. The whole point of the `Context` design pattern is to add and change parameters without the user having to adjust their class, while the same user code would be executed.

But that's the opposite of what we want - if you ignore any provided piece of info, you are likely to be clobbering other data. We actually want any new or changed parameters to be a new code path that is unsupported until the user gets around to implementing it.

So the API for right now should reflect these properties:

 - there are just two cases
 - in both cases you really need to use all the information
 - you should know at construction time whether the policy supports the case you need

Something like:

```java
abstract class FileNamePolicy implements Serializable {

  abstract String filenameForGlobalData(int shardNum, int numShards);

  String filenameForWindowedData(
      int shardNum, int numShards, BoundedWindow w, PaneInfo paneInfo) {
    throw new UnsupportedOperationException(
        "This filename policy cannot be used with windowed input");
  }

  // And maybe this
  abstract boolean verifySupportsWindowedOutput;
}
```

This could conceivably be factored into some base classes named for what they support, or even separate abstract classes, either of which could be provided to `FileBasedSink`, to make the distinction easy to detect.

It would be best to actually detect whether it supports windowing or not, but I don't know if we have the bandwidth to get something together along those lines.